### PR TITLE
Extended generics support

### PR DESCRIPTION
**Pull request by FabianMartin**
**Describe the PR**
Extends the generic support for go 1.8

1. Adds support for nested generics (generic[string, generic[int]])
2. Adds support for array parameter on generics (generic[string, []string])
3. Adds support for generic properties on structs
4. Adds support for built-in types in generics

**Relation issue**
https://github.com/swaggo/swag/issues/1170
https://github.com/swaggo/swag/issues/1213

**Additional context**
All changes are backward compatible and have been tested with Go 1.15, 1.16, 1.17 and 1.18.4. It is possible that other issues, besides the ones mentioned, have been fixed.

### CODE REVIEW

File: generics.go [hunk 1]
```diff
@@ -101,18 +157,106 @@ func (pkgDefs *PackagesDefinitions) parametrizeStruct(original *TypeSpecDef, ful
...
-			return genTypeSpec.TypeSpec.Type
+			if genTypeSpec.ArrayDepth > 0 {
+				genTypeSpec.ArrayDepth--
+				return &ast.ArrayType{Elt: resolveType(expr, field, genericParamTypeDefs)}
+			}
+			return genTypeSpec.Type()
 		}
 	} else if asArray, ok := expr.(*ast.ArrayType); ok {
 		return &ast.ArrayType{Elt: resolveType(asArray.Elt, field, genericParamTypeDefs), Len: asArray.Len, Lbrack: asArray.Lbrack}
 	}
```
Comment (ubogdan) [inline generics.go:215]: Else if is hard to follow, please rewrite it with a `switch case` statement for better readability.

```go
	switch astExpr := expr.(type) {
	case *ast.Ident:
		//TODO fill the code here
	case *ast.ArrayType:
		return &ast.ArrayType{
			Elt:    resolveType(astExpr.Elt, field, genericParamTypeDefs),
			Len:    astExpr.Len,
			Lbrack: astExpr.Lbrack,
		}
	}
```

File: generics.go [hunk 2]
```diff
@@ -26,29 +51,47 @@ func typeSpecFullName(typeSpecDef *TypeSpecDef) string {
...
 
 	for i, genericParam := range genericParams {
-		tdef, ok := pkgDefs.uniqueDefinitions[genericParam]
-		if !ok {
-			return nil
+		arrayDepth := 0
+		for {
+			var isArray = len(genericParam) > 2 && genericParam[:2] == "[]"
+			if isArray {
+				genericParam = genericParam[2:]
```
Comment (ubogdan) [inline generics.go:75]: isArray is used only once. Please put the excretion inside if statement

```go
if ! (len(genericParam) > 2 && genericParam[:2] == "[]") {
 break 
}
genericParam = genericParam[2:]
arrayDepth++
```

File: generics.go [hunk 3]
```diff
@@ -26,29 +51,47 @@ func typeSpecFullName(typeSpecDef *TypeSpecDef) string {
...
+			}
 		}
 
-		genericParamTypeDefs[original.TypeSpec.TypeParams.List[i].Names[0].Name] = tdef
+		tdef := pkgDefs.FindTypeSpec(genericParam, original.File, parseDependency)
+		if tdef == nil {
+			genericParamTypeDefs[original.TypeSpec.TypeParams.List[i].Names[0].Name] = &genericTypeSpec{
+				ArrayDepth: arrayDepth,
+				TypeSpec:   nil,
+				Name:       genericParam,
```
Comment (ubogdan) [inline generics.go:87]: Don't need to initialize TypeSpec since is nill by default.

**Review by ubogdan — CHANGES_REQUESTED** [review]
Good job so far. Please update according to the comments.

**Comment by FabianMartin** [comment]
Hi @ubogdan,
thanks for the feedback. I have made the suggested changes and added new tests.

**Review by ubogdan — APPROVED** [review]
LGTM

**Comment by ubogdan** [comment]
@FabianMartin Thanks for your contribution.

**Comment by johan-lejdung** [comment]
Fantastic, much appreciated @FabianMartin ! Looking forward to this getting released so that I can try it with my generic structs that's currently not being picked up correctly 🙏 @ubogdan do you have any publicly available release schedule?

With any luck this might also solve https://github.com/swaggo/swag/issues/1253

**Comment by FabianMartin** [comment]
@johan-lejdung Thank you for your feedback. A fix for your issue is included in PR https://github.com/swaggo/swag/pull/1281.
